### ISSUE TITLE
[continuous-integration] fuzz failure in `Ip6::InsertMplOption`

### ISSUE THREAD

**Issue by Irving-cl** [openthread/openthread#5460]
**Describe the bug**
Some PRs now get failure in fuzzers: #5455 #5458 #5459 

I think the reason is that, in `src/core/net/ip6.cpp:283`, it's called like this:
```
    aMessage.CopyTo(8, 0, hbhLength, aMessage);
```
This message copies to itself.

And in the newly updated `CopyTo` method, it uses `memcpy`, when `hbhLength` is greater than `8`, it would cause a 'overlap', which causes the error indicated in fuzzer:

> ==11==ERROR: AddressSanitizer: memcpy-param-overlap: memory ranges [0x00000131bf10,0x00000131bf20) and [0x00000131bf18, 0x00000131bf28) overlap
> SCARINESS: 10 (memcpy-param-overlap)
>     #0 0x51fce4 in __asan_memcpy /src/llvm-project/compiler-rt/lib/asan/asan_interceptors_memintrinsics.cpp:22:3
>     #1 0x55d010 in ot::Message::Write(unsigned short, unsigned short, void const*) /src/openthread/src/core/common/message.cpp:540:9
>     #2 0x55e066 in ot::Message::CopyTo(unsigned short, unsigned short, unsigned short, ot::Message&) const /src/openthread/src/core/common/message.cpp:555:18
>     #3 0x592bf0 in ot::Ip6::Ip6::InsertMplOption(ot::Message&, ot::Ip6::Header&, ot::Ip6::MessageInfo&) /src/openthread/src/core/net/ip6.cpp:283:22
>     #4 0x5958fc in ot::Ip6::Ip6::SendRaw(ot::Message&) /src/openthread/src/core/net/ip6.cpp:1108:9
>     #5 0x554960 in otIp6Send /src/openthread/src/core/api/ip6_api.cpp:167:37
>     #6 0x553e7b in LLVMFuzzerTestOneInput /src/openthread/tests/fuzz/ip6_send.cpp:72:13
>     #7 0x458621 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:559:15
>     #8 0x457d65 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:471:3
>     #9 0x459d07 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::__Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:771:7
>     #10 0x459f09 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:800:3
>     #11 0x449bd3 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:851:6
>     #12 0x471642 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
>     #13 0x7fd1a8e8c83f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2083f)
>     #14 0x41e568 in _start (out/ip6-send-fuzzer+0x41e568)

This error blocks some PRs now. Shall we use `memmove`? Or we should change the implementation of `Ip6::InsertMplOption` about the copy usage.

@abtink

**Comment by abtink** [comment]
I think using `memmove` in `Write()` in this case should help address this. Will submit a PR for this soon.

I think the overlap write was an issue with original implementation as well and was being masked by the fact `CopyTo` would read content into a separate 16 bytes buffer from source message and the write them in destination. If there was an overlap it could overwrite the same data.

**Comment by abtink** [comment]
Unfortunately using `memmove` instead of `memcpy` won't work in this case.
We can write into a next chunk in destination message before we get the chance to read it.

**Comment by abtink** [comment]
Thanks @Irving-cl  for submitting this and finding the issue.

I submitted https://github.com/openthread/openthread/pull/5462 to unblock.

As I mentioned in the PR commit message, the `CopyTo()` (even with `memmove()`) cannot be used to copy bytes forward within same message (we don't seem to use it anywhere in the code to forward copy within same message and I added an assert for now to block such a use).

The bytes in a message are stored in bunch of `Buffer` chunks. When a forward copy spans over multiple `Buffer` chunks, the `CopyTo()` implementation can potentially write stuff to a next `Buffer` chunk before getting the chance to read it. One way I can think of to handle this, is to go through the `Buffer` chunks in backward order (though not easy to realize as the `Buffer`s are in singly linked list).
